File: cpp/leetcode/408.cpp
class Solution {
public:
    bool validWordAbbreviation(string word, string abbr) {
        if(word.empty() || abbr.empty()) return word == abbr;
        int i = 0, j = 0;
        while(i < word.size() && j < abbr.size()){
            if(isalpha(abbr[j])){
                if(abbr[j] == word[i]) {j++; i++;}
                else return false;
            } else {
                if (abbr[j] == '0') return false; // impossible abbr
                int tmp = 0;
                while(j < abbr.size() && isdigit(abbr[j]))  tmp = 10 * tmp + abbr[j++] - '0';
                i += tmp;
            }
        }
        return i == word.size() && j == abbr.size();
    }
};

File: python/leetcode/408.py
class Solution(object):
    def validWordAbbreviation(self, word, abbr):
        i, j = 0, 0
        while i < len(abbr) and j < len(word):
            if abbr[i] == word[j]:
                i += 1
                j += 1
            elif abbr[i].isdigit() and abbr[i]!='0':
                k = i
                while k <len(abbr) and abbr[k].isdigit(): k += 1
                j += int(abbr[i:k])
                i = k
            else:
                return False
        return i == len(abbr) and j == len(word)